### PR TITLE
feat: implement insert/update/delete builders

### PR DESCRIPTION
**Pull request by lundibundi**

### CODE REVIEW

**Comment by tshemsedinov** [comment]
looks like you forgot to execute
```
npm run fmt
```

File: lib/select-builder.js [hunk 1]
```diff
@@ -2,9 +2,9 @@
 
 const { iter } = require('@metarhia/common');
 
-const { QueryBuilder, makeParamValue } = require('./query-builder');
-const { ConditionsBuilder } = require('./conditions-builder');
-const { mapJoinIterable } = require('./utils');
+const { makeParamValue } = require('./query-builder');
+const { checkTypeOrQuery, mapJoinIterable } = require('./utils');
+const { QueryConditionsBuilder } = require('./query-conditions-builder');
```
Comment (tshemsedinov) [inline lib/select-builder.js]: I'd prefer to add `.js`

File: lib/select-builder.js [hunk 2]
```diff
@@ -274,15 +119,21 @@ class SelectBuilder extends QueryBuilder {
...
+    return typeof subquery === 'function'
+      ? subquery(new SelectBuilder(this.params, this.options))
+      : subquery;
+  }
+
+  _processFrom(from) {
     return mapJoinIterable(
       from,
-      f => (typeof f === 'string' ? f : `${f.table} AS ${f.alias}`),
+      f => (!f.alias ? String(f.table) : `${f.table} AS ${f.alias}`),
```
Comment (tshemsedinov) [inline lib/select-builder.js]: Don't we use arrow-parens: always ?

File: lib/update-builder.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  _processSet(set) {
+    return mapJoinIterable(
+      set,
+      s => `${s.column} = ${makeParamValue(s.value, this.params)}`,
+      ', '
+    );
+  }
+
+  build() {
+    let query = 'UPDATE';
```
Comment (tshemsedinov) [inline lib/update-builder.js]: Maybe: query = []; .push, push, .join ?

File: lib/utils.js [hunk 4]
```diff
@@ -30,9 +31,19 @@ const joinIterable = (val, sep) => {
   return res;
 };
 
+const checkTypeOrQuery = (value, name, type) => {
+  if (!(value instanceof QueryBuilder) && typeof value !== type) {
```
Comment (tshemsedinov) [inline lib/utils.js]: It seems to me we need `OR` logic here: `!instance || !type`

File: lib/delete-builder.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+const { QueryConditionsBuilder } = require('./query-conditions-builder');
+const { SelectBuilder } = require('./select-builder');
+
+class DeleteBuilder extends QueryConditionsBuilder {
+  constructor(params, options) {
+    super(params, options);
+    this.operations = {
+      from: null,
+    };
```
Comment (tshemsedinov) [inline lib/delete-builder.js]:
```suggestion
    this.operations = { from: null };
```

Comment (tshemsedinov) [reply lib/delete-builder.js]: In line 16 we have one-liner

File: lib/delete-builder.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  _whereValueMapper(subquery) {
+    return typeof subquery === 'function'
+      ? subquery(new SelectBuilder(this.params, this.options))
+      : subquery;
+  }
+
+  build() {
+    let query = 'DELETE';
```
Comment (tshemsedinov) [inline lib/delete-builder.js]: Maybe: query = []; .push, push, .join ?

File: lib/delete-builder.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    const from = this.operations.from;
```
Comment (tshemsedinov) [inline lib/delete-builder.js]:
```suggestion
    const { from } = this.operations;
```

File: lib/delete-builder.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    if (!from) {
+      throw new Error('Cannot generate SQL, tableName is not defined');
+    }
+    query += String(from.table) + (from.alias ? `AS ${from.alias}` : '');
```
Comment (tshemsedinov) [inline lib/delete-builder.js]: Looks like `from.table` will be string already
```suggestion
    const alias = from.alias ? `AS ${from.alias}` : '';
    query += `$from.table $alias`;
```

File: lib/select-builder.js (hunk 2 above)
Comment (lundibundi) [reply lib/select-builder.js]: Oops, looks like i mixed up styles. Will update.

File: lib/utils.js (hunk 4 above)
Comment (lundibundi) [reply lib/utils.js]: No, this is "exit condition" for error cases.

File: lib/update-builder.js (hunk 3 above)
Comment (lundibundi) [reply lib/update-builder.js]: I prefer to avoid array-join unnecessarily.

File: lib/delete-builder.js (hunk 8 above)
Comment (lundibundi) [reply lib/delete-builder.js]: Done.

File: lib/delete-builder.js (hunk 7 above)
Comment (lundibundi) [reply lib/delete-builder.js]: Done.

File: lib/delete-builder.js (hunk 5 above)
Comment (lundibundi) [reply lib/delete-builder.js]: Done.

File: lib/select-builder.js (hunk 1 above)
Comment (lundibundi) [reply lib/select-builder.js]: Done.

**Review by tshemsedinov — APPROVED** [review]
LGTM

**Pull request merged**